File: 0096-unique-binary-search-trees/0096-unique-binary-search-trees.py
class Solution:
  def numTrees(self, n: int) -> int:
    dp = dict()
    dp[0] = 1
    
    def helper(seg):
      total = 0
      for i in range(1, seg+1):
        total += dp[i-1]*dp[seg-i]
      dp[seg] = total
    
    for i in range(1, n+1):
      helper(i)
    return dp[n]

File: 0096-unique-binary-search-trees/0096-unique-binary-search-trees.java
class Solution {
  public void helper(int seg, Map<Integer, Integer> dp) {
    int total = 0;
    for (int n = 1; n <= seg; n++) {
      total += dp.get(n - 1) * dp.get(seg - n);
    }
    dp.put(seg, total);
  }
  
  public int numTrees(int n) {
    Map<Integer, Integer> dp = new HashMap<Integer, Integer>();
    dp.put(0, 1);
    for (int i = 1; i <= n; i++) helper(i, dp);
    return dp.get(n);
  }
}